consensus/src/test/kotlin/maru/consensus/state/StateTransitionImplTest.kt

```diff
@@ -0,0 +1,113 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+
+class StateTransitionImplTest {
+  private val currentBlock = DataGenerators.randomBeaconBlock(9u)
+  private val preState =
+    DataGenerators.randomBeaconState(9u).copy(
+      latestBeaconBlockHeader = currentBlock.beaconBlockHeader,
+      latestBeaconBlockRoot = HashUtil.bodyRoot(currentBlock.beaconBlockBody),
+    )
+
+  private val newBeaconBlockBody = DataGenerators.randomBeaconBlockBody()
+  private val stateRootBlockHeader =
+    DataGenerators.randomBeaconBlockHeader(10u).copy(
+      parentRoot = currentBlock.beaconBlockHeader.hash,
+      stateRoot = ByteArray(0),
+      bodyRoot = HashUtil.bodyRoot(newBeaconBlockBody),
+    )
+  private val stateRootHash =
+    HashUtil.stateRoot(
+      preState.copy(
+        latestBeaconBlockHeader = stateRootBlockHeader,
+        latestBeaconBlockRoot = stateRootBlockHeader.bodyRoot,
+      ),
+    )
+  private val newBlockHeader = stateRootBlockHeader.copy(stateRoot = stateRootHash)
+  private val newBlock = BeaconBlock(newBlockHeader, newBeaconBlockBody)
+  private val validatorProvider =
+    object : ValidatorProvider {
+      override fun getValidatorsForBlock(header: BeaconBlockHeader): SafeFuture<Set<Validator>> =
+        SafeFuture.completedFuture(setOf(newBlockHeader.proposer))
+    }
+
+  private val proposerSelector =
+    object : ProposerSelector {
+      override fun getProposerForBlock(header: BeaconBlockHeader): SafeFuture<Validator> =
+        SafeFuture.completedFuture(newBlockHeader.proposer)
+    }
+
+  @Test
+  fun `processBlock should return failure if block validation fails`() {
+    val blockValidator =
+      BlockValidator { block, _, prevBlockHeader, _ ->
+        if (block == newBlock && prevBlockHeader == preState.latestBeaconBlockHeader) {
+          SafeFuture.completedFuture(Err(BlockValidator.BlockValidationError("Block validation failed")))
+        } else {
+          SafeFuture.completedFuture(Ok(true))
+        }
+      }
+    val stateTransition =
+      StateTransitionImpl(
+        blockValidator = blockValidator,
+        validatorProvider = validatorProvider,
+        proposerSelector = proposerSelector,
+      )
+
+    val result = stateTransition.processBlock(preState, newBlock).get()
+    assertThat(result is Err).isTrue()
+    assertThat(result.component2()).isNotNull()
+    assertThat(result.component2()?.message).isEqualTo("State Transition failed. Reason: Block validation failed")
+  }
+
+  @Test
+  fun `processBlock should return ok if all checks pass`() {
+    val blockValidator =
+      BlockValidator { block, _, prevBlockHeader, _ ->
+        if (block == newBlock && prevBlockHeader == preState.latestBeaconBlockHeader) {
+          SafeFuture.completedFuture(Ok(true))
+        } else {
+          SafeFuture.completedFuture(Err(BlockValidator.BlockValidationError("Block validation failed")))
```

Comment:
I don't see what's the point in adding a failure case here if it's a precondition of a test that validation is successful. If it's not - that's another test